### PR TITLE
Fix bug with multiple regions in PCA and snp_allele_counts

### LINKED ISSUE #266: Better caching for Ag3.pca()
Currently the Ag3.snp_allele_counts() function does a pretty good job of saving results and retrieving from the disk cache, even if the sample query parameter has been cosmetically changed but still specifies the same samples. This is because the sample query is resolved to a list of sample indices [here](https://github.com/malariagen/malariagen-data-python/blob/226931beeef14697c3956f8fc79b315ce911b1ee/malariagen_data/ag3.py#L5666).

The same could be done for the Ag3.pca() function to ensure better cache retrieval.

### LINKED ISSUE #336: Anopheles PCA cannot handle list of regions
E.g.:

```
ag3.pca(
    n_snps=100_000,
    region=['3R', '3L'],
    sample_sets=releases,
    sample_query=(
        f'country in {west_countries} and '
        f'aim_species not in ["arabiensis", "intermediate_gambcolu_arabiensis"]'
    )
)
```

...generates:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[28], line 1
----> 1 west_gambcolu_pca, west_gambcolu_evr = ag3.pca(
      2     n_snps=100_000,
      3     region=['3R', '3L'],
      4     sample_sets=releases,
      5     sample_query=(
      6         f'country in {west_countries} and '
      7         f'aim_species not in ["arabiensis", "intermediate_gambcolu_arabiensis"]'
      8     )
      9 )

File /home/conda/global/446e7c4fbc16c08647a3c0fa672fd8c760afcebbda20837e64f1a10aed61c0ec-20221209-184039-583571-2-binder-4.1.0/lib/python3.9/site-packages/malariagen_data/anopheles.py:3290, in AnophelesDataResource.pca(self, region, n_snps, thin_offset, sample_sets, sample_query, site_mask, min_minor_ac, max_missing_an, n_components)
   3286     site_mask = self._default_site_mask
   3288 debug("normalize params for consistent hash value")
   3289 params = dict(
-> 3290     region=self.resolve_region(region).to_dict(),
   3291     n_snps=n_snps,
   3292     thin_offset=thin_offset,
   3293     sample_sets=self._prep_sample_sets_arg(sample_sets=sample_sets),
   3294     sample_query=sample_query,
   3295     site_mask=site_mask,
   3296     min_minor_ac=min_minor_ac,
   3297     max_missing_an=max_missing_an,
   3298     n_components=n_components,
   3299 )
   3301 debug("try to retrieve results from the cache")
   3302 try:

AttributeError: 'list' object has no attribute 'to_dict'
```

Affects `pca()` and `snp_allele_counts()`.

### PR DESCRIPTION
**Pull request by alimanfoo**
* Resolves #336 
* Also resolves #266 along the way
* Also upgrades pre-commit hooks because of install problems on CI

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -1,7 +1,12 @@
...
+    # whitespace before ':' - doesn't work well with black
+    E203
+    # module level import not at top of file
+    E402
+    # line too long - let black worry about that
+    E501
+    # do not assign a lambda expression, use a def
+    E731
+    # line break before binary operator
+    W503
```
Comment (alimanfoo) [inline .flake8:12]: This is needed because of upgrading flake8 in pre-commit config.

File: malariagen_data/anopheles.py [hunk 2]
```diff
@@ -4494,7 +4519,7 @@ def _read_cohort_metadata(self, *, sample_set):
                     },
                     inplace=True,
                 )
-            except FileNotFoundError:
+            except Exception:
```
Comment (alimanfoo) [inline malariagen_data/anopheles.py:4522]: This exception seems to have changed, needed to broaden.

**Review by leehart — APPROVED** [review]
LGTM.

Thanks for the helpful inline comments @alimanfoo . How do you do that?

**Comment by alimanfoo** [comment]
> Thanks for the helpful inline comments @alimanfoo . How do you do that?

No worries - I just go into the files tab and add single comments, a bit like doing a code review on my own PR.

**Pull request merged**